pkg/webhook/server.go

```diff
@@ -0,0 +1,104 @@
+package webhook
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/openservicemesh/osm/pkg/certificate"
+	"github.com/openservicemesh/osm/pkg/constants"
+	"github.com/openservicemesh/osm/pkg/metricsstore"
+)
+
+// CertRotatedFunc is a callback to perform other actions when the server's HTTPS cert gets rotated.
+type CertRotatedFunc func(cert *certificate.Certificate) error
+
+// Server is a construct to run generic HTTPS webhook servers.
+type Server struct {
+	name         string
+	cm           *certificate.Manager
+	server       *http.Server
+	onCertChange CertRotatedFunc
+
+	mu   sync.Mutex
+	cert tls.Certificate
+}
+
+// NewServer returns a new server based on the input. Run() must be called to start the server.
+func NewServer(name, namespace string, port int, cm *certificate.Manager, handlers map[string]http.HandlerFunc, onCertChange CertRotatedFunc) (*Server, error) {
+	mux := http.NewServeMux()
+
+	for path, h := range handlers {
+		mux.Handle(path, metricsstore.AddHTTPMetrics(h))
+	}
+
+	s := &Server{
+		name:         name,
+		cm:           cm,
+		onCertChange: onCertChange,
+	}
+	s.server = &http.Server{
+		Addr:    fmt.Sprintf(":%d", port),
+		Handler: mux,
+		// #nosec G402
+		TLSConfig: &tls.Config{
+			GetCertificate: func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+				s.mu.Lock()
+				defer s.mu.Unlock()
+				return &s.cert, nil
+			},
+			MinVersion: tls.VersionTLS13,
+		},
+	}
+	// set the certificate once, which will also call onCertChange.
+	if err := s.setCert(name, namespace); err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+// Run actually starts the server. It blocks until the passed in context is done.
+func (s *Server) Run(ctx context.Context) {
+	log.Info().Msgf("Starting %s webhook server on: %s", s.name, s.server.Addr)
+	go func() {
+		err := s.server.ListenAndServeTLS("", "") // err is always non-nil
+		log.Error().Err(err).Msgf("%s webhook HTTP server shutdown")
+	}()
+
+	// Wait on exit signals
+	<-ctx.Done()
+
+	// Stop the servers
+	if err := s.server.Shutdown(ctx); err != nil {
```

Comment:
You're passing a canceled context into `Shutdown` here; I'm not sure what the implications of that are.

Reply:
good catch! fixed